src/replication.c

```diff
@@ -3444,161 +3476,227 @@ static client *findReplica(char *host, int port) {
     return NULL;
 }
 
-/* FAILOVERTO HOST PORT [TIMEOUT <timeout>] */
-void failovertoCommand(client *c) {
-    if (server.cluster_enabled) {
-        addReplyError(c,"FAILOVERTO not allowed in cluster mode. "
-                        "Use CLUSTER FAILOVER command instead.");
-        return;
+const char *getFailoverStateString() {
+    switch(server.failover_state) {
+        case NO_FAILOVER: return "no-failover";
+        case FAILOVER_IN_PROGRESS: return "failover-in-progress";
+        case FAILOVER_WAIT_FOR_SYNC: return "waiting-for-sync";
+        default: return "unknown";
     }
+}
 
-    if (server.masterhost != NULL) {
-        addReplyError(c,"FAILOVERTO is not valid when server is a replica.");
-        return;
-    }
+/* Resets the internal failover configuration, this needs
+ * to be called after a failover either succeeds or fails
+ * as it includes the client unpause. */
+void clearFailoverState() {
+    server.failover_end_time = 0;
+    server.force_failover = 0;
+    zfree(server.target_replica_host);
+    server.target_replica_host = NULL;
+    server.target_replica_port = 0;
+    server.failover_state = NO_FAILOVER;
+    unpauseClients();
+}
+
+/* Abort an ongoing failover if one is going on. */
+void abortFailover(const char *err) {
+    if (server.failover_state == NO_FAILOVER) return;
 
-    if (server.failover_end_time) {
-        addReplyError(c, "FAILOVERTO already in progress.");
+    if (server.target_replica_host) {
+        serverLog(LL_NOTICE,"FAILOVER to %s:%d aborted: %s",
+            server.target_replica_host,server.target_replica_port,err);  
+    } else {
+        serverLog(LL_NOTICE,"FAILOVER to ANY ONE aborted: %s",err);  
+    }
+    if (server.failover_state == FAILOVER_IN_PROGRESS) {
+        replicationUnsetMaster();
+    }
+    clearFailoverState();
+}
+
+/* 
+ * FAILOVER TO <HOST> <IP> [FORCE] [TIMEOUT <timeout>] 
```

Comment:
i think the TO argument can be optional, this way we don't need the ANY ONE trick.
```suggestion
 * FAILOVER [TO <HOST> <IP> [FORCE] ] [TIMEOUT <timeout>] 
```

Reply:
I was trying to document it like it was one command with two subcommands. Otherwise it would look like this:
```
FAILOVER [TO <HOST> <IP> [FORCE]] [TIMEOUT <timeout>] [ABORT]
```
Which I find much harder to follow.

Reply:
ok. got it.. but if it's a sub-command, i'm still not sure the `TO` is a good sub-command.
maybe the sub-command should be something like `START`, and `[TO <host> <port>]` would still be an optional argument.

Reply:
Sorry, I now understand what you are getting at, "FAILOVER" by itself would pick a replica and failover. I still like the current wording, I'll address all the other comments and then ping the group about what the command will be named like.

Reply:
yes.. the main "advantage" is that TO becomes optional and we don't need to use the "ANY ONE" trick